src/edu/csus/ecs/pc2/clics/API202306/CLICSContestAccess.java

```diff
@@ -0,0 +1,167 @@
+            cap.add(API_CAPABILITY_CONTEST_THAW);
+        }
+        if(SubmissionService.isTeamSubmitAllowed(sc)) {
+            cap.add(API_CAPABILITY_TEAM_SUBMIT);
+        }
+        if(ClarificationService.isTeamSubmitClarificationAllowed(sc)) {
+            cap.add(API_CAPABILITY_TEAM_CLAR);
+        }
+        if(SubmissionService.isProxySubmitAllowed(sc)) {
+            cap.add(API_CAPABILITY_PROXY_SUBMIT);
+        }
+        if(ClarificationService.isProxySubmitClarificationAllowed(sc)) {
+            cap.add(API_CAPABILITY_PROXY_CLAR);
+        }
+        if(SubmissionService.isAdminSubmitAllowed(sc)) {
+            cap.add(API_CAPABILITY_ADMIN_SUBMIT);
+        }
+        if(ClarificationService.isAdminSubmitClarificationAllowed(sc)) {
+            cap.add(API_CAPABILITY_ADMIN_CLAR);
+        }
+        if(AwardService.isAwardsUpdateAllowed(sc)) {
+            cap.add(API_CAPABILITY_AWARDS_UPDATE);
+        }
+        capabilities = cap.toArray(new String[0]);
+
+        /*
+         * As per the spec, here are the possible endpoints.
+         * These should match the class names in this package, eg. Contest -> ContestService, Run -> RunService, etc.
+         */
+        String [] serviceNames = { "Contest", "JudgementType", "Language", "Problem", "Group", "Organization",
+                "Team", "Person", "Account", "State", "Submission", "Judgement", "Run", "Clarification",
+                "Scoreboard", "EventFeed", "Award", "Commentary"
+        };
+
+        ArrayList<CLICSEndpoint> epList = new ArrayList<CLICSEndpoint>();
+
+        for (String service: serviceNames) {
+            if(!invokeServiceEndpointAccessMethod(epList, service, sc)) {
+                if(Utilities.isDebugMode()) {
+                    System.err.println("No endpoint access for service " + service);
+                }
+                controller.getLog().log(Log.WARNING, "No endpoint access for service " + service);
+            }
+        }
+        if(!epList.isEmpty()) {
+            endpoints = epList.toArray(new CLICSEndpoint[0]);
+        }
+    }
+
+    /**
+     * Create a fully qualified class name for the supplied web service endpoint.
+     * eg. Group would create: edu.csus.ecs.pc2clics.API202306.GroupService
+     * Then call its getEndpointProperties(SecurityContext) static method to see what properties the
+     * service supports for the supplied SecurityContext (user).
+     * Note: If Java reflection makes you queasy, please stop here and just skip reading the code in this method.
+     *
+     * @param epList List to add the CLICSEndpointObject to, if one was returned
+     * @param serviceName The service name only (excluding the trailing "Service")
+     * @param sc User's security context
+     * @return true if this endpoint is accessible by the user in the SecurityContext
+     */
+    private boolean invokeServiceEndpointAccessMethod(ArrayList<CLICSEndpoint> epList, String serviceName, SecurityContext sc) {
```

Comment:
I think the name of this method is confusing -- it sounds like an imperative, like "I'm going to invoke the Service Endpoint".   Since what the method is doing is determining whether the specified service is allowed to be invoked given the specified SecurityContext, it seems like a better name for the method would be (something like) `canInvokeServiceEndpoint()`  or maybe `isAllowedToInvoke...`.

Also, I was taught that boolean functions with side-effects (in this case, updating the epList) are kind of not a good idea in general software engineering terms (although I grant that the alternative -- a classwide variable being updated -- is probably just as bad so ignore this unless you have a better fix...)

Reply:
Actually, it DOES invoke the Service's endpoint **_Access Method_** to determine what properties exist and are accessible.  It does NOT invoke the **_ServiceEndpoint_**.  The only confusion may be that it only invokes the SEAM if it exists, which is a small nit to pick.